intersection and union types with java and scala (#299)
ScottLogic/blog: _posts/2025-02-21-intersection-and-union-types-with-java-and-scala.md

---
title: Intersection and Union types with Java and Scala
date: 2025-02-21 00:00:00 Z
categories:
- Tech
tags:
- Java, Scala
author: magnussmith
summary: In this post we will see  how Intersection types help us better model type constraints promoting reuse and how Union types increase code flexibility.  We will compare and contrast approaches and how to use in the latest Java and Scala.    
image: magnussmith/assets/java.jpg
---

![intersection_union.webp]({{site.baseurl}}/magnussmith/assets/intersection_union.webp)

This is the third post in a series exploring types and type systems.   Previous posts have looked at

- [Algebraic Data Types with Java]({{site.baseurl}}/2025/01/20/algebraic-data-types-with-java.html)

- [Variance, Phantom and Existential types in Java and Scala]({{site.baseurl}}/2025/02/17/variance-in-java-and-scala.html)

## Intersection and Union Types with Java and Scala

One of the difficult things for modern programming languages to get right is around providing flexibility when it comes to expressing complex relationships. As languages evolve, they need to give us tools to model the nuances of real world problems more accurately in our code.

> Modern software systems are much more complex than in years gone by, and developers need type systems that can accurately express intricate relationships. **Intersection types** and **union types** provide this expressiveness by allowing programmers to specify more precise type constraints.

Reviewer: ```suggestion
Modern software systems are much more complex than they were in years gone by, and developers need type systems that can accurately express intricate relationships. **Intersection types** and **union types** provide this expressiveness by allowing programmers to specify more precise type constraints.
```

> We will see how these powerful features of the type system, while have been present in some form, are now becoming increasingly refined in languages like Java and Scala, providing benefits that help us write more expressive, safer, and ultimately more maintainable code.

Reviewer: ```suggestion
We will see how these powerful features of the type system, while having been present in some form, are now becoming increasingly refined in languages like Java and Scala, providing benefits that help us write more expressive, safer, and ultimately more maintainable code.
```

> In this post, we'll dive deep into these concepts, exploring how they're implemented in the latest versions of **Java 23** and **Scala 3.6**. We'll see practical examples, compare and contrast the approaches, identify where they shine and where they fall short we will discuss how to bridge the gaps. 

Reviewer: ```suggestion
In this post, we'll dive deep into these concepts, exploring how they're implemented in the latest versions of **Java 23** and **Scala 3.6**. We'll see practical examples, compare and contrast the approaches, identify where they shine, and where they fall short we will discuss how to bridge the gaps. 
```

> ### A brief historical timeline.

Reviewer: ```suggestion
### A Brief Historical Timeline.
```

> Before we go into detail of their application lets take a brief look at their historical context.

Reviewer: ```suggestion
Before we go into detail about their application, let's take a brief look at their historical context.
```

> - **1930s-1960s**: The theoretical foundations for intersection and union types can be traced back to type theory and mathematical logic, specifically in areas dealing with type polymorphism and subtyping. Thinkers like Haskell Curry, Alonzo Church, and later researchers in the field of type systems laid the groundwork for understanding types not just as mathematical sets, but as structured entities with relationships.

Reviewer: ```suggestion
- **1930s-1960s**: The theoretical foundations for intersection and union types can be traced back to type theory and mathematical logic, specifically in areas dealing with type polymorphism and subtyping. Thinkers like Haskell Curry, Alonzo Church, and later researchers in the field of type systems have laid the groundwork for understanding types not just as mathematical sets, but as structured entities with relationships.
```

> - **1970s-1980s**: Early type systems (ML, Lisp) lacked explicit intersection/union types. C and Pascal introduced unions.

Reviewer: Not exactly a must have but I feel like C's unions deserve a callout for not being as type safe as the modern things you're talking about? I always saw them as more of a way to conserve memory, and practically mostly saw them (ab)used to treat one type as another (a Matrix as float[4][4] + float[16] was a pretty common pattern).

- **1990s**: Functional languages (ML, OCaml) formalised **[Abstract Data Types (ADTs)]({{site.baseurl}}/2025/01/20/algebraic-data-types-with-java.html)**, enabling safer unions.
- **2000s**: **Scala** (2004) introduced explicit intersection type with `with` syntax.

> - **2010**s: [**TypeScript**](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html) (2012) embraced and popularized union (`A | B`) and intersection (`A & B`) types in mainstream development as essential features for adding static typing on top of dynamic foundations. 

Reviewer: ```suggestion
- **2010**s: [**TypeScript**](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html) (2012) embraced and popularised union (`A | B`) and intersection (`A & B`) types in mainstream development as essential features for adding static typing on top of dynamic foundations. 
```
Blog prefers British English :)

    resource.log("Closing resource (Java).");
    resource.close();
  }
}
~~~~

This ensures that manageResource (and manageResourceJava) only accepts objects that guarantee *both* closing and logging capabilities, enhancing safety and clarity.

**The Diamond Problem? Not with Intersections!**

A classic problem in multiple inheritance is the "diamond problem." This arises when a class inherits from two classes that share a common ancestor, and both parent classes override a method from that ancestor. Which version of the method should the inheriting class use?

Intersection types inherently sidestep the diamond problem. When you create an intersection type like `Resettable & Configurable`, you are *not* creating a new inheritance hierarchy. Instead, you are defining a *constraint*. In Java any type that satisfies `Resettable & Configurable` must simply *implement* both `Resettable` and `Configurable`. There's no shared ancestor in the traditional inheritance sense, and thus no ambiguity about which method to call.  In essence, intersection types are about *composition* of capabilities, not hierarchical inheritance, which naturally avoids the diamond problem scenarios.

> In Scala where traits include the method definition then the right implementation of a method that appears in both traits is resolved through a process called `linearisation`.  When a method is called, Scala searches this linearised order from right to left to find the first implementation of the method. So in the case of Scala order matters for intersection types.

Reviewer: ```suggestion
In Scala, where traits include the method definition, then the right implementation of a method that appears in both traits is resolved through a process called `linearisation`.  When a method is called, Scala searches this linearised order from right to left to find the first implementation of the method. So, in the case of Scala, order matters for intersection types.
```

**Java: Simulating Union Types with Sealed Classes/Interfaces - Pattern Matching to the Rescue!**

Let's update the Java union type simulation, using pattern matching for instanceof to simplify the handleInputJava method and make it more readable:

Java

~~~~ java
sealed interface InputType {}
record StringInput(String value) implements InputType {}
record IntInput(Integer value) implements InputType {}

class SimulatedUnionExample {
  static void handleInputJava(InputType input) {
    var result =
        switch (input) {
          case StringInput stringInput -> "Received string input (Java): " + stringInput.value();
          case IntInput intInput -> "Received int input (Java): " + intInput.value();
        };  // Compiler knows all subtypes are handled due to 'sealed'.
    System.out.println(result);
  }
  public static void main(String[] args) {
    handleInputJava(new StringInput("Hello Java"));
    handleInputJava(new IntInput(456));
  }
}
~~~~

> While the underlying structure with sealed interfaces remains, the handleInputJava method is now cleaner thanks to pattern matching checks. This makes working with simulated union types in Java significantly less verbose compared to older versions.

Reviewer: ```suggestion
While the underlying structure with sealed interfaces remains, the handleInputJava method has now become cleaner thanks to pattern matching checks. This has made working with simulated union types in Java significantly less verbose compared to older versions.
```

record Failure<S, E>(E error) implements ResultJava<S, E> {}

class ErrorHandlingJavaExample {
  
  public static void main(String[] args) {
    System.out.println(evaluate(divideJava(10, 2)));
    System.out.println(evaluate(divideJava(5, 0)));
  }
  
  static ResultJava<Integer, String> divideJava(int a, int b) {
    if (b == 0) return failure("Division by zero! (Java)");
    else return success(a / b);
  }

  static <S, E> String evaluate(ResultJava<S, E> result) {
    return switch (result) {
      case Success<S, E> success -> "Result (Java): " + success.value();
      case Failure<S, E> failure -> "Error (Java): " + failure.error();
    };
  }

  static <S, E> ResultJava<S, E> success(S value) {
    return new Success<>(value);
  }

  static <S, E> ResultJava<S, E> failure(E error) {
    return new Failure<>(error);
  }
}
~~~~

### **Gaps and Simulation: Java's Union Type Story**

As we’ve discussed, Java simulates union types through sealed hierarchies, while Scala provides native union types. Let's delve deeper into the implications and differences:

**Exhaustiveness Checking in Pattern Matching: Java vs. Scala**

> When dealing with unions, both simulated and real it is very likely that we are going to undertake pattern matching.

Reviewer: ```suggestion
When dealing with unions, both simulated and real, it is very likely that we are going to undertake pattern matching.
```

A key advantage of Java's simulation of union types using sealed interfaces is compile-time exhaustiveness checking when used with pattern matching.

> Because sealed hierarchies have a fixed, known set of subtypes, the Java compiler can verify that your pattern matching logic handles all possible cases. If you miss a case in a switch expression over a sealed type the Java compiler will issue a compile-time error.  This dramatically increases safety and helps prevent bugs where you forget to handle a possible variant of your union type.

Reviewer: ```suggestion
Because sealed hierarchies have a fixed, known set of subtypes, the Java compiler can verify that your pattern matching logic handles all possible cases. If you miss a case in a switch expression over a sealed type, the Java compiler will issue a compile-time error.  This dramatically increases safety and helps prevent bugs where you forget to handle a possible variant of your union type.
```

**Scala 3.6's Union Types and Non-Exhaustive Matching (by Default)**

In contrast, while Scala 3.6's pattern matching on native union types is very flexible and powerful, it does not, by default, enforce exhaustiveness in the same way.  If you pattern match on a union type and miss a case, the Scala compiler might issue a _warning_ depending on settings, but it's not the same level of enforced exhaustiveness as Java provides with sealed types.

> ###### A couple of practical steps to help mitigate this  are:

Reviewer: ```suggestion
###### A couple of practical steps to help mitigate this issue are:
```
